### DISCUSSION TITLE
Restreaming Solution

### DISCUSSION THREAD

**Discussion by PanicKk** [savonet/liquidsoap#3359] [Q&A]
> Is there a way to change the input source on the fly?

# Use Case
I am trying to implement a **.NET application** that allows users to re-stream content to their desired **RTMP Servers**. Such thing is possible with **FFMPEG** via **OS pipes**, but that approach is not really reliable in my case.

I saw that **liquidsoap** offers the possibility to mix sources or add a fallback source in case the main source has issues.

I've been trying for the past 5 hours to write a basic script that takes an RTMP input as the main source and has an MP4 file as fallback source, but the RTMP source is not playing at all.

It works fine with MP4 files, but I need a way to make it work for RTMP inputs, if that's possible.

How can I pass an RTMP Stream as input for the main source / stream ? Can I change that input on the fly? 

Can another RTMP Stream be set as a fallback source? I believe I would be able to change the fallback source. In that case I would be able to switch between live-streams.

### Current Script
```
settings.init.force_start.set(true)
settings.server.telnet.set(true)

#source_1 = input.harbor('rtmp://server/live/key, port=1935)
#source_2 = input.harbor('rtmp://server/live/key', port=2935)

source_1 = single('/home/panickk/Videos/testVideo.mp4')
source_2 = single('/home/panickk/Videos/3min_timer.mp4')

mixed = add([source_1,source_2])

streams = ref([])
count = ref(0)

encoder = %ffmpeg(format="flv", %video(codec="libx264", width=1280, height=720, pixel_format="yuv420p"), %audio(codec="aac"))

def create_stream(url) =
  if list.assoc.mem(url, !streams) then
    "Stream for url #{url} already exists!"
  else
    out = output.url(
      id="restream-#{!count}",
      fallible=true,
      url=url,
      encoder,
      mixed
    )
    count := !count + 1
    streams := [...!streams, (url, out.shutdown)]
    "OK!"
  end
end

def delete_stream(url) = 
  if not list.assoc.mem(url, !streams) then
    "Stream for url #{url} does not exist!"
  else
    shutdown = list.assoc(url, !streams)
    shutdown()
    streams := list.filter((fun (el) -> fst(el) != url), !streams)
    "OK!"
  end
end


server.register(namespace="restream",
                description="Redirect a stream.",
                usage="start <url>",
                "start",
                create_stream)
server.register(namespace="restream",
                description="Stop a dynamic playlist.",
                usage="stop <url>",
                "stop",
                delete_stream)
```

## Requirements
A simple re-stream is not hard to make, just push any content via FFMPG to the RTMP Server of choice, but switching between content is a must. Something like how OBS allows you to switch between scenes seamlessly, I want to implement something that allows users to change the content they are re-streaming on the fly (example: vid1 -> vid2 - without losing connection to RTMP Server)

**Comment by vitoyucepi** [comment]
Hi @PanicKk,
I think you're looking for the [`switch`](https://www.liquidsoap.info/doc-2.2.0/reference.html#switch) and [`fallback`](https://www.liquidsoap.info/doc-2.2.0/reference.html#fallback) functions.
They are working similarly. The only exception is that `switch` supports custom conditions.
Code samples from [the book](https://www.liquidsoap.info/book/book.pdf) can help you implement the desired pipeline for your streaming service. 

If no RTMP streams are available, a default video can be casted. Here's an example.
```ruby
s1 = input.rtmp('rtmp://server/live/key')
backup = single('/home/panickk/Videos/testVideo.mp4')

s = fallback([s1, backup], track_sensitive=false)
```

If you wish to specify switching conditions that are not standard, here is an example of how to use the `switch` function.
```ruby
s1 = input.rtmp('rtmp://server/live/key1')
s2 = input.rtmp('rtmp://server/live/key2')

use_s1 = interactive.bool('Activate stream 1', false)
use_s2 = interactive.bool('Activate stream 2', false)

s = switch([(use_s1, s1), (use_s2, s2)], track_sensitive=false)
```

I am using `interactive.bool` as an example of an externally controlled variable. If needed, you can exchange data using JSON over http or telnet.

**Reply by PanicKk** [reply]
I tried using input.RTMP but it does not work for some reason. It cannot open connection to RTMP Server.

![image](https://github.com/savonet/liquidsoap/assets/60987556/f85c8d21-4819-4320-ae0e-9ee330e106c7)

**Reply by vitoyucepi** [reply]
It's possible that I don't understand your problem correctly.
I suggest checking [my answer in another discussion](https://github.com/savonet/liquidsoap/discussions/3167#discussioncomment-6259066) where I tested various input methods.

**Reply by PanicKk** [reply]
I was trying with v2.1.3 and got the mentioned errors. I tried installing the latest v2.2.0 via OPAM and now I get different errors:
![image](https://github.com/savonet/liquidsoap/assets/60987556/c0504bf6-f7bf-4af8-a258-f2b23c2fa40c)
![image](https://github.com/savonet/liquidsoap/assets/60987556/66d5a664-46ad-4fbe-afbd-0e3bf4b89c35)

**Reply by vitoyucepi** [reply]
Sorry, but could you upload a part of your script responsible for receiving the RTMP stream?
Anyway, could you test this script and tell if it's working or not?
Keep in mind that the server is running on localhost, so *127.0.0.1* can be used as its IP address.
1. `main.liq`
   ```ruby
   s = input.rtmp("rtmp://0.0.0.0:8000/a/b")
   enc = %ffmpeg(format="mpegts", %video(codec="libx264"))
   output.harbor(enc, mksafe(s), port=8001, mount="test")
   ```
1. `liquidsoap main.liq`
1. `vlc http://127.0.0.1:8001/test`
   Should show a black screen.
1. Configure streaming in OBS.
   ![4](https://github.com/savonet/liquidsoap/assets/35035879/8897f6be-7f28-4dc0-aba0-a1c2ab208358)
1. Start streaming from OBS.


You have to install the `ffmpeg` package from opam to use `input.rtmp`.
It'll be fine if you post code snippets instead of the terminal screenshots.
If you're familiar with docker, you can use [`savonet/liquidsoap`](https://hub.docker.com/r/savonet/liquidsoap) container images.

**Reply by PanicKk** [reply]
I only have Nginx with the RTMP module running in a Docker container, no particular script for any processing, just basic RTMP Push.

I did install ffmpeg from opam, along with all the other dependencies, but still no luck.

I'm willing to give a shot at the Docker image, but there is no info at all on how to use in on the Docker Hub. Do you mind to help me with that?

**Reply by PanicKk** [reply]
I found a way to run it from the Liquidsoap CookBook.

`docker run -it -v /home/panickk/Videos/tmp1:/home/tmp1 --entrypoint /bin/bash --network ls-test savonet/liquidsoap:main`

The tmp1 directory contains the script file, and I set the network with the intention to let the container communicate with the Nginx container, but it can not.

Here is the current script:
```
source_1 = input.rtmp("rtmp://127.0.0.1:1935/live/test")
source_2 = single('/home/tmp1/3min_timer.mp4')

encoder = %ffmpeg(format="flv", %video(codec="libx264", width=1280, height=720, pixel_format="yuv420p"), %audio(codec="aac"))

stream = fallback(track_sensitive=false, [source_1,source_2])

out = output.url(
  id="restream-#{!count}",
  fallible=true,
  url="rtmp://server/live/key",
  encoder,
  stream
)
```

Here is what I get when I run the script:

```
liquidsoap@af8959998e3f:/home/tmp1$ liquidsoap ls2.liq 
2023/08/29 11:27:15 >>> LOG START
2023/08/29 11:27:14 [main:3] Liquidsoap 2.1.0+git@4c4e684a
2023/08/29 11:27:14 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.5.0 sedlex=2.4 menhirLib=20210419 curl=0.9.1 dtools=0.4.4 duppy=0.9.2 cry=0.6.5 mm=0.7.3 xmlplaylist=0.1.5 lastfm=0.3.3 ogg=0.7.0 ogg.decoder=0.7.0 vorbis=0.8.0 vorbis.decoder=0.8.0 opus=0.2.1 opus.decoder=0.2.1 speex=0.4.1 speex.decoder=0.4.1 mad=0.5.1 flac=0.3.0 flac.ogg=0.3.0 flac.decoder=0.3.0 dynlink=[distributed with Ocaml] lame=0.3.5 shine=0.2.2 frei0r=0.1.2 fdkaac=0.3.2 theora=0.4.0 theora.decoder=0.4.0 ffmpeg=1.1.2 bjack=0.1.6 alsa=0.3.0 ao=0.2.4 samplerate=0.1.6 taglib=0.3.8 ssl=0.5.9 magic=0.7.3 camomile=1.0.2 inotify=2.3 faad=0.5.0 soundtouch=0.1.9 portaudio=0.2.3 pulseaudio=0.1.5 ladspa=0.2.2 dssi=0.1.5 tsdl=v0.9.8 tsdl-image=0.3.2 camlimages=4.2.6 cohttp-lwt-unix=4.0.0 prometheus-app=1.1 srt.constants=0.2.2 srt.types=0.2.2 srt.stubs=0.2.2 srt.stubs.locked=0.2.2 srt=0.2.2 lo=0.2.0 gd=1.0a5
2023/08/29 11:27:14 [main:3] 
2023/08/29 11:27:14 [main:3] DISCLAIMER: This version of Liquidsoap has been
2023/08/29 11:27:14 [main:3] compiled from a snapshot of the development code.
2023/08/29 11:27:14 [main:3] As such, it should not be used in production
2023/08/29 11:27:14 [main:3] unless you know what you are doing!
2023/08/29 11:27:14 [main:3] 
2023/08/29 11:27:14 [main:3] We are, however, very interested in any feedback
2023/08/29 11:27:14 [main:3] about our development code and committed to fix
2023/08/29 11:27:14 [main:3] issues as soon as possible.
2023/08/29 11:27:14 [main:3] 
2023/08/29 11:27:14 [main:3] If you are interested in collaborating to
2023/08/29 11:27:14 [main:3] the development of Liquidsoap, feel free to
2023/08/29 11:27:14 [main:3] drop us a mail at <savonet-devl@lists.sf.net>
2023/08/29 11:27:14 [main:3] or to join the slack chat at <http://slack.liquidsoap.info>.
2023/08/29 11:27:14 [main:3] 
2023/08/29 11:27:14 [main:3] Please send any bug report or feature request
2023/08/29 11:27:14 [main:3] at <https://github.com/savonet/liquidsoap/issues>.
2023/08/29 11:27:14 [main:3] 
2023/08/29 11:27:14 [main:3] We hope you enjoy this snapshot build of Liquidsoap!
2023/08/29 11:27:14 [main:3] 
2023/08/29 11:27:14 [clock:3] Using builtin (low-precision) implementation for latency control
2023/08/29 11:27:15 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2023/08/29 11:27:15 [frame:3] Video frame size set to: 1280x720
2023/08/29 11:27:15 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2023/08/29 11:27:15 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2023/08/29 11:27:15 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2023/08/29 11:27:15 [sandbox:3] Sandboxing disabled
2023/08/29 11:27:15 [video.converter:3] Using preferred video converter: ffmpeg.
2023/08/29 11:27:15 [audio.converter:3] Using samplerate converter: ffmpeg.
2023/08/29 11:27:15 [single_0:3] "/home/tmp1/3min_timer.mp4" is static, resolving once for all...
2023/08/29 11:27:16 [clock.main:3] Streaming loop starts in auto-sync mode
2023/08/29 11:27:16 [clock.main:3] Delegating synchronisation to CPU clock
2023/08/29 11:27:16 [single_0:3] Prepared "/home/tmp1/3min_timer.mp4" (RID 0).
2023/08/29 11:27:16 [switch_0:3] Switch to single_0.
^C2023/08/29 11:27:26 [main:3] Shutdown started!
2023/08/29 11:27:26 [main:3] Waiting for main threads to terminate...
2023/08/29 11:27:26 [clock.main:3] Streaming loop stopped.
2023/08/29 11:27:26 [main:3] Main threads terminated.
2023/08/29 11:27:26 [threads:3] Shutting down scheduler...
[rtmp @ 0x7fa5fc0058c0] Cannot open connection tcp://127.0.0.1:1935?listen&listen_timeout=-1000
2023/08/29 11:27:26 [threads:3] Scheduler shut down.
2023/08/29 11:27:26 [main:3] Cleaning downloaded files...
[aac @ 0x7fa614053b40] 2 frames left in the queue on closing
2023/08/29 11:27:26 [main:3] Freeing memory...
```

I also tried your suggested script, I get the same connection error. What am I missing here.

Note: Testing this time was done via docker image

**Reply by vitoyucepi** [reply]
To debug the local scripts, I run a docker container like this: `docker run -it --rm --network host -v /tmp/test1:/tmp/test1 -w /tmp/test1 --entrypoint bash savonet/liquidsoap:v2.2.0`. It provides an interactive shell so I can run `liquidsoap main.liq`.
I use `--network host` to forward connections as is.

In production, I will embed all the required scripts into the image. Instead of a host network, I'll publish all the desired ports.

Anyway, you already have opam and liquidsoap installed.
Could you run this script and share the output?
```ruby
s = input.rtmp("rtmp://0.0.0.0:8000/a/b")
enc = %ffmpeg(format="mpegts", %video(codec="libx264"))
output.harbor(enc, mksafe(s), port=8001, mount="test")
```
The error ``
this value has no method `rtmp`
`` caused because `ffmpeg` is not installed. Installing `ffmpeg` from opam allows liquidsoap to use `input.rtmp`.

**Reply by PanicKk** [reply]
I ran this script inside the docker image and this is what I get:

```
liquidsoap@docker-desktop:/home/tmp1$ liquidsoap test.liq
2023/08/29 11:35:24 >>> LOG START
2023/08/29 11:35:23 [main:3] Liquidsoap 2.1.0+git@4c4e684a
2023/08/29 11:35:23 [main:3] Using: bytes=[distributed with OCaml 4.02 or above] pcre=7.5.0 sedlex=2.4 menhirLib=20210419 curl=0.9.1 dtools=0.4.4 duppy=0.9.2 cry=0.6.5 mm=0.7.3 xmlplaylist=0.1.5 lastfm=0.3.3 ogg=0.7.0 ogg.decoder=0.7.0 vorbis=0.8.0 vorbis.decoder=0.8.0 opus=0.2.1 opus.decoder=0.2.1 speex=0.4.1 speex.decoder=0.4.1 mad=0.5.1 flac=0.3.0 flac.ogg=0.3.0 flac.decoder=0.3.0 dynlink=[distributed with Ocaml] lame=0.3.5 shine=0.2.2 frei0r=0.1.2 fdkaac=0.3.2 theora=0.4.0 theora.decoder=0.4.0 ffmpeg=1.1.2 bjack=0.1.6 alsa=0.3.0 ao=0.2.4 samplerate=0.1.6 taglib=0.3.8 ssl=0.5.9 magic=0.7.3 camomile=1.0.2 inotify=2.3 faad=0.5.0 soundtouch=0.1.9 portaudio=0.2.3 pulseaudio=0.1.5 ladspa=0.2.2 dssi=0.1.5 tsdl=v0.9.8 tsdl-image=0.3.2 camlimages=4.2.6 cohttp-lwt-unix=4.0.0 prometheus-app=1.1 srt.constants=0.2.2 srt.types=0.2.2 srt.stubs=0.2.2 srt.stubs.locked=0.2.2 srt=0.2.2 lo=0.2.0 gd=1.0a5
2023/08/29 11:35:23 [main:3] 
2023/08/29 11:35:23 [main:3] DISCLAIMER: This version of Liquidsoap has been
2023/08/29 11:35:23 [main:3] compiled from a snapshot of the development code.
2023/08/29 11:35:23 [main:3] As such, it should not be used in production
2023/08/29 11:35:23 [main:3] unless you know what you are doing!
2023/08/29 11:35:23 [main:3] 
2023/08/29 11:35:23 [main:3] We are, however, very interested in any feedback
2023/08/29 11:35:23 [main:3] about our development code and committed to fix
2023/08/29 11:35:23 [main:3] issues as soon as possible.
2023/08/29 11:35:23 [main:3] 
2023/08/29 11:35:23 [main:3] If you are interested in collaborating to
2023/08/29 11:35:23 [main:3] the development of Liquidsoap, feel free to
2023/08/29 11:35:23 [main:3] drop us a mail at <savonet-devl@lists.sf.net>
2023/08/29 11:35:23 [main:3] or to join the slack chat at <http://slack.liquidsoap.info>.
2023/08/29 11:35:23 [main:3] 
2023/08/29 11:35:23 [main:3] Please send any bug report or feature request
2023/08/29 11:35:23 [main:3] at <https://github.com/savonet/liquidsoap/issues>.
2023/08/29 11:35:23 [main:3] 
2023/08/29 11:35:23 [main:3] We hope you enjoy this snapshot build of Liquidsoap!
2023/08/29 11:35:23 [main:3] 
2023/08/29 11:35:23 [clock:3] Using builtin (low-precision) implementation for latency control
2023/08/29 11:35:24 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2023/08/29 11:35:24 [frame:3] Video frame size set to: 1280x720
2023/08/29 11:35:24 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2023/08/29 11:35:24 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2023/08/29 11:35:24 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2023/08/29 11:35:24 [sandbox:3] Sandboxing disabled
2023/08/29 11:35:24 [video.converter:3] Using preferred video converter: ffmpeg.
2023/08/29 11:35:24 [audio.converter:3] Using samplerate converter: ffmpeg.
2023/08/29 11:35:24 [harbor:3] Adding handler for 'GET /test' on port 8001
2023/08/29 11:35:24 [clock.main:3] Streaming loop starts in auto-sync mode
2023/08/29 11:35:24 [clock.main:3] Delegating synchronisation to CPU clock
2023/08/29 11:35:24 [mksafe:3] Switch to safe_blank.
^C2023/08/29 11:35:28 [main:3] Shutdown started!
2023/08/29 11:35:28 [main:3] Waiting for main threads to terminate...
2023/08/29 11:35:28 [harbor:3] Removing handler for 'GET /test' on port 8001
2023/08/29 11:35:28 [clock.main:3] Streaming loop stopped.
2023/08/29 11:35:28 [main:3] Main threads terminated.
2023/08/29 11:35:28 [threads:3] Shutting down scheduler...
[rtmp @ 0x7ff2c4005880] Cannot open connection tcp://0.0.0.0:8000?listen&listen_timeout=-1000
2023/08/29 11:35:28 [threads:3] Scheduler shut down.
2023/08/29 11:35:28 [main:3] Cleaning downloaded files...
2023/08/29 11:35:28 [main:3] Freeing memory...
```

I am not unable to stream to that 'server' from OBS.

**Reply by vitoyucepi** [reply]
Please use `savonet/liquidsoap:v2.2.0`.
Everything looks like it's working correctly.
It's possible that you can't connect to a server because you didn't publish the ports. Use `--network host` or configure the ports.

Oh, and this one `input.rtmp("rtmp://127.0.0.1:1935/live/test")`, replace `127.0.0.1` with `0.0.0.0`.

**Reply by PanicKk** [reply]
This is driving me crazy....

I tried with host network but it throws the same connection error. I cannot configure ports, I'm assuming that's because port 1935 is already occupied by nginx container, or is there some other ports I need to configure?

Here is my step-by-step workflow:
1. Run Nginx with the RTMP module: `docker run -d --name nginx_live -p 1935:1935 -p 35000:35000 nginx_live`
2. Run the Docker Image: `docker run -it -v /home/panickk/Videos/tmp1:/home/tmp1 --network host --entrypoint /bin/bash savonet/liquidsoap:v2.2.0`
3. Execute the liquidsoap script:
```
source_1 = input.rtmp("rtmp://0.0.0.0:1935/live/test")
source_2 = single('/home/tmp1/3min_timer.mp4')

encoder = %ffmpeg(format="flv", %video(codec="libx264", width=1280, height=720, pixel_format="yuv420p"), %audio(codec="aac"))

stream = fallback(track_sensitive=false, [source_1,source_2])

out = output.url(
  id="restream",
  fallible=true,
  url="rtmp://server/live/key,
  encoder,
  stream
)
```

I get:
```
[rtmp @ 0x7f40fd828100] Cannot open connection tcp://0.0.0.0:1935?listen&listen_timeout=-1000&tcp_nodelay=0
2023/08/29 12:08:21 [input.ffmpeg:2] Decoding failed: Avutil.Error(Address already in use)
```

What address is already in use? The RTMP input? What does it matter? OBS is the only service using it, and it only Pushes the stream, that should not stop it from consuming, right?

I don't understand why this just won't work. Is there something I am missing?

**Reply by vitoyucepi** [reply]
It appears that the port is currently occupied, which is why you are unable to listen with another process.
Why wouldn't you change the port? `rtmp://0.0.0.0:8000/live/test` or something.
Also, you could check the port by executing `ss -ltpn` to list all TCP port listeners with the corresponding process names.

**Reply by PanicKk** [reply]
I am pushing content to that port, it makes sense to then fetch that content as the input in the script. I don't see why would that be an issue...

**Reply by vitoyucepi** [reply]
1. `docker run -d --name nginx_live -p 1935:1935 -p 35000:35000 nginx_live`
   You occupy TCP ports 1935 and 35000 on the host.
1. `docker run -it -v /home/panickk/Videos/tmp1:/home/tmp1 --network host --entrypoint /bin/bash savonet/liquidsoap:v2.2.0`
   You can't use any of the ports that are already busy. So ports 1935 and 35000 are out.

**Reply by toots** [reply]
@PanicKk `lsof -lntp` should tell you the list of opened ports and which application is using it.

**Reply by PanicKk** [reply]
I figured it out. There has been a huge misunderstanding!

The issue was that `input.rtmp` somehow creates a new RTMP instance and it expects content to be pushed there.

I thought that it only accepts an already existing RTMP Server and only fetches the content from that source. So, I was pushing content to an Nginx Server with the RTMP Module and was passing that URL as an input. That is what caused the `port already in use` issues.

It was not mentioned anywhere how that process is handled, or at least I didn't find any, and sadly that caused all of this trouble.

Is there a way to pass RTMP Source as the Input, not creating a new one? Because for my intentions, this is very resource consuming...

I need to fetch from two RTMP Sources, one being the main stream and the other the fallback / backup one. So when the mainstream stops, the fallback comes in play. The fallback source has to be changed on the fly, using http or telnet, so I gotta figure that out too.

Thank you guys, for your help! I really appreciate it.

**Reply by vitoyucepi** [reply]
@PanicKk,
To have it work the desired way, set `listen=false` and provide the server's address. Additionally check the [`input.rtmp`](https://www.liquidsoap.info/doc-2.2.0/reference.html#input.rtmp) documentation.
But do you have to use the nginx rtmp module if you can send incoming rtmp connections directly to liquidsoap?

**Reply by PanicKk** [reply]
Yes, I already have an RTMP Server solution for another part of my software, so I want to reuse the input from that server and re-stream it to other servers.

The attribute `listen=false` does the job for me. What is left is to learn how to change the value of the defined sources, so the re-stream content can change based on user preferences, and add the possibility to change it via http requests or telnet.

Is there any particular source or tutorial that teaches the liquidsoap scripting language, besides the official book?

It is really time consuming to scrape the whole book / documentation, if there is a faster way to get my hands-on.

**Reply by vitoyucepi** [reply]
I've learned the language by
1. reading the [book](https://www.liquidsoap.info/book/book.pdf);
1. reading the [documentation](https://www.liquidsoap.info/doc-2.2.0/);
1. reading the [lib sources](https://github.com/savonet/liquidsoap/tree/v2.2.0/src/libs);
1. reading the [tests](https://github.com/savonet/liquidsoap/tree/v2.2.0/tests/language);
1. doing the experiments;
1. asking the questions;
1. answering the questions.

**Comment by Mikulskii** [comment]
I'm sorry to interfere, perhaps not on business.
But, it seemed to me that the topic is somewhat similar to mine:
I stream a video playlist 24/7 and want to switch to "live" from time to time via OBS. `switch` and `fallback` operators work as expected. 
But! as soon as I stop the stream in OBS, the broadcast and Liquidsoap freezes. 
And rtmp doesn't even switch to safe blank. It only helps to return playback if you turn on OBS again. As if Liquidsoap decides that `input.rtmp` should be permanently active.
Maybe there are ideas or tips on how to forcibly stop/break/forget connection of the rtmp source?
```
interactive.harbor(port=8010, uri="/control")

rtmp = mksafe(input.rtmp(listen=false, "rtmp://localhost/test"))
videos = mksafe(playlist("f:/VODs/NEW_PLAYLIST"))
dummy = playlist(loop=false, "e:/liquidsoap/media")
p = interactive.bool("r1", false)
videos = switch(track_sensitive=false, transition_length=0., [(p, rtmp), ({true}, videos)])

url = "rtmp://localhost/live"
enc = %ffmpeg(format="flv",
%video(codec="libx264", width=1280, height=720, pixel_format="yuv420p",
b="3500k", preset="ultrafast", bf=2, framerate=30, g=60),
%audio(codec="libmp3lame", samplerate=44100, b="256k"))

output.url(fallible=true, url=url, enc, videos)

```

**Reply by PanicKk** [reply]
Hi @Mikulskii. What version of liquidsoap are you using? I am pretty new to liquidsoap myself, but I don't have this issue.

I have two RTMP inputs, one as the main stream and one as a fallback, no matter how many times I stop the main stream on OBS, it falls back to fallback input always.

Here is my basic script:
```
source_1 = input.rtmp("rtmp://server:1935/live/key1", listen=false)
source_2 = input.rtmp("rtmp://server:1935/live/key2", listen=false)

encoder = %ffmpeg(format="flv", %video(codec="libx264", width=1280, height=720, pixel_format="yuv420p"), %audio(codec="aac"))

stream = fallback(track_sensitive=false, [source_1,source_2])

out = output.url(
  id="restream",
  fallible=true,
  url="rtmp://server:1935/live/key3",
  encoder,
  stream
)
```

**Reply by Moonbase59** [reply]
Hey @Mikulskii, I’m actually doing a videostream, using a cam input harbor to facilitate cam/OBS streaming with immediate failover. Maybe have a look here for part of the code:

https://github.com/savonet/liquidsoap/discussions/3370

You don’t have to use extra authentication and connect/disconnect scripting as my example does. For a totally simple test you can
```ruby
# set up an input for studio cams/OBS
# port 8005, user=source, pw=hackme
cam = input.harbor("cam")
```
and _ffmpeg_ (as shown) or OBS should be able to stream to that.

Only don’t leave this out in the wild after testing, or malicious people will take over your video stream. ;-)

**Reply by Mikulskii** [reply]
@PanicKk Oh... I tried on 2.1.4 and on 2.2.0 with the same result (at the same time I tried your method again). Since I'm practically a noob, I first test everything on Windows, and then I upload it to Ubuntu VPS.
Maybe it's just worth trying on a VPS... Thanks for the answer!

@Moonbase59 Thanks! I will definitely try to experiment with your code!
And yes, I will be careful :D

**Reply by PanicKk** [reply]
@Mikulskii I first tried it on Ubuntu, since I use it as my daily driver, but I was facing issues. I believe that it was due to the dependencies of Liquidsoap not being the right version or something. I thought to give it a try with the docker image of Liquidsoap, and everything worked fine.

Here is the command I use for the docker image:
`docker run -p 8000:8000 -p 8003:8003 -it -v /home/panickk/ls_scripts:/home/ls_scripts --entrypoint /bin/bash savonet/liquidsoap:v2.2.0`

The `/home/panickk/ls_scripts` is where I have the liquid soap scripts, so I mount that volume to the container since the image does not have nano or vim in it and I didn't bother to install them. The ports are just there, not functional atm, but I plan to use them later.

**Reply by Mikulskii** [reply]
@PanicKk , @Moonbase59 Hey guys, thanks a lot for your answers <3
But everything turned out to be quite simple: it was necessary to add the line `play_restart on;` to the nginx-rtmp-module config for `application`...
after that everything began to work as expected ;)

**Reply by zachryit** [reply]
@Moonbase59 please can you demo how to connect the cam via OBS?

`# set up an input for studio cams/OBS
# port 8005, user=source, pw=hackme
cam = input.harbor("cam")`

**Comment by PanicKk** [comment]
Hi! I've been able to solve most of the issues I've been facing, thanks to this discussion. What I need to figure out now, is if there is a way to change the source of an `input.rtmp` even though it is already connected to an output.

I thought of having two `input.rtmp`s and switch between those. The switch works fine and all, but after some time, I may want to change the source of the first input and then switch to that ( same for the second one ) while always keeping a connection to the output server without interruption between source switch.

Demonstration:
```

# The source of each input is an RTMP Server (Nginx with the RTMP module)
source_one = input.rtmp("rtmp://server:1935/live/key", listen=false)
source_two = input.rtmp("rtmp://server:1935/live/different-key", listen=false)

stream = switch(track_sensitive=false, [(booean1, source_one), (boolean2, source_two)]

# So after some time I decide to switch to source_two for a while, and later switch to source_one but with a different stream url
source_one = input.rtmp("rtmp://server:1935/live/new-key", listen=false)

# ... switch back and worth, with the same / different inputs, as desired
```

I tried using references:
```
input1 = ref("rtmp://server:1935/live/key)

input1 := "rtmp://server:1935/live/new-key"
```

But, even though the url changes, the input sources is still the same, even after the switch.

Does it save the url in a cache / buffer once the output starts streaming? Is there a way to do that sort of switch / manipulation?

Maybe shutdown the first input, change the url, then start again?

I would really appreciate your help. If that is possible ofc...
